csharp/src/Ice/LocatorInfo.cs

```diff
@@ -212,16 +213,31 @@ internal void ClearCache(Reference reference)
             return (endpoints, cached);
         }
 
-        internal async ValueTask<ILocatorRegistryPrx?> GetLocatorRegistryAsync(CancellationToken cancel)
+        internal async Task<ILocatorRegistryPrx?> GetLocatorRegistryAsync(CancellationToken cancel)
         {
-            if (_locatorRegistry == null)
+            lock (_mutex)
             {
-                ILocatorRegistryPrx? registry = await Locator.GetRegistryAsync(cancel: cancel).ConfigureAwait(false);
+                if (_locatorRegistryRetrieved)
```

Comment:
The difficulty with the "await the same task" pattern is cancellation. What if the first caller cancels?
With the "let all the calls run" pattern, we don't have this issue.

Reply:
In this case where `GetRegistryAsync` is cheap I think it's fine to make multiple calls. Otherwise, we could do the same thing as locator calls where we perform a "singed-shared" remote call which is never canceled and each caller await on `task.WaitAsync(cancel)` instead.